### ISSUE TITLE
'Type is already specialized' after narrowing generic with type var in function scope

### ISSUE THREAD

**Issue by peterschutt** [microsoft/pyright#4755]
While building some unittests I came across a difference between mypy and pyright when a generic type is narrowed with a type var within a function scope.

```python
from __future__ import annotations

from typing import Generic, TypeVar

T = TypeVar("T")

class MyGeneric(Generic[T]):
    ...

def fn() -> None:
    t2 = TypeVar("t2")
    still_generic = MyGeneric[t2]
    concrete = still_generic[str]
    reveal_type(concrete)

```

mypy reports:

```
(.venv) peter@peter-Inspiron-15-5510:~/PycharmProjects/pythonProject7$ mypy .
main.py:14: note: Revealed type is "def () -> main.MyGeneric[builtins.str]"
Success: no issues found in 1 source file
```

pyright reports

```
(.venv) peter@peter-Inspiron-15-5510:~/PycharmProjects/pythonProject7$ pyright .
No configuration file found.
No pyproject.toml file found.
stubPath /home/peter/PycharmProjects/pythonProject7/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.298
/home/peter/PycharmProjects/pythonProject7/main.py
  /home/peter/PycharmProjects/pythonProject7/main.py:13:16 - error: Type "MyGeneric[t2]" is already specialized (reportGeneralTypeIssues)
  /home/peter/PycharmProjects/pythonProject7/main.py:14:17 - information: Type of "concrete" is "Type[MyGeneric[t2]]"
1 error, 0 warnings, 1 information 
Completed in 0.618sec
```

It doesn't matter if the 2nd type var is defined inside, or outside of the function.

**Comment by erictraut** [comment]
Pyright's behavior is correct here. The expression `still_generic = MyGeneric[t2]` is defining a variable assignment. It looks like you are expecting it to be a type alias, but type aliases cannot be defined in functions. They are allowed only at the module level or within classes. PEPs 484 and 613 weren't clear about this, but it was discussed among the Python type checker maintainers, and this was the consensus of the typing community. See [this thread](https://mail.python.org/archives/list/typing-sig@python.org/thread/CGOO7GPPECGMLFDUDXSSXTRADI4BXYCS/#WSWVBF76Q3WQKXWXWLLDLGRUMB3Z2TSS) for details.

As you're probably aware, the traditional way to define a type alias in Python is syntactically indistinguishable from a regular variable assignment. Type checkers must use heuristics to distinguish between the two. In your code, pyright is treating the statement as a variable assignment because it is located within a function.

If you use PEP 613 and annotate the statement as an explicit `TypeAlias`, pyright will emit an error indicating that a type alias cannot be defined within a function.

![image](https://user-images.githubusercontent.com/7040122/224062995-69e2b11e-abd9-4381-9d06-e13bdcf20927.png)

So, the correct way to code this is to move the TypeVar and type alias definitions out of the function, like this:

```python
from typing import Generic, TypeAlias, TypeVar

T = TypeVar("T")

class MyGeneric(Generic[T]):
    ...

T2 = TypeVar("T2")
still_generic: TypeAlias = MyGeneric[T2]

def fn() -> None:
    concrete = still_generic[str]
    reveal_type(concrete)
```

**Comment by peterschutt** [comment]
Thanks @erictraut - I really appreciate the level of detail in your response, and for linking that mailing group thread.

All the best.
